### PR TITLE
Update the way to fetch DailyHunt pagination, and apply paging library to news

### LINKED ISSUE #4712: [News] Dailyhunt - change next page mechanism
**Describe the bug**
Based on Dailyhunt's feedback on next page mechanism:
“For populating the next page, you should use the nextPageUrl fetched as a part of the current response. NextPageUrl contains parameters to take care of the content user has already seen and the same will not be repeated. I suppose currently, you are just updating the pageNumber to fetch new content. This is resulting in repetition of items in the same feed.”

We should change the "scroll to next page" mechanism to avoid repetition.

### PR DESCRIPTION
**Pull request by proevan**
to fix #4712, and apply paging library to news

### CODE REVIEW

File: app/src/main/java/org/mozilla/rocket/adapter/DelegatePagedListAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package org.mozilla.rocket.adapter
+
+import android.view.ViewGroup
+import androidx.paging.PagedListAdapter
+import androidx.recyclerview.widget.DiffUtil
+
+open class DelegatePagedListAdapter(
+    private val delegatesManager: AdapterDelegatesManager,
+    diffCallback: DiffUtil.ItemCallback<DelegateAdapter.UiModel>
+) : PagedListAdapter<DelegateAdapter.UiModel, DelegateAdapter.ViewHolder>(diffCallback) {
```
Comment (proevan) [inline app/src/main/java/org/mozilla/rocket/adapter/DelegatePagedListAdapter.kt:10]: for extending `PagedListAdapter`

File: app/src/main/java/org/mozilla/rocket/content/news/data/NewsDataSourceFactory.kt [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    private val appContext: Context
+) : DataSource.Factory<Int, NewsItem>() {
+
+    lateinit var category: String
+    lateinit var language: String
+
+    override fun create(): DataSource<Int, NewsItem> {
+        val dailyHuntProvider = DailyHuntProvider.getProvider(appContext)
+        return DailyHuntNewsRemoteDataSource(appContext, dailyHuntProvider, category, language)
+        // TODO: support other sources
```
Comment (proevan) [inline app/src/main/java/org/mozilla/rocket/content/news/data/NewsDataSourceFactory.kt:18]: support other sources in later commits

File: app/src/main/java/org/mozilla/rocket/content/news/data/NewsDataSourceFactory.kt [hunk 3]
```diff
@@ -1,24 +1,35 @@
...
         } else {
-            NewsRepository(RssNewsRemoteDataSource(newsProvider))
+            RssNewsRemoteDataSource(newsProvider, category)
         }
     }
+
+    sealed class PageKey {
+        class PageNumberKey(val number: Int) : PageKey()
+        class PageUrlKey(val url: String) : PageKey()
+    }
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/rocket/content/news/data/NewsDataSourceFactory.kt:34]: It's nice you have tow types of keys

File: app/src/main/java/org/mozilla/rocket/content/news/data/dailyhunt/DailyHuntNewsRemoteDataSource.kt [hunk 4]
```diff
@@ -1,90 +1,177 @@
...
+    }
+
+    override fun loadBefore(params: LoadParams<PageKey>, callback: LoadCallback<PageKey, NewsItem>) {
+        // Do nothing
+    }
+
+    override fun loadAfter(params: LoadParams<PageKey>, callback: LoadCallback<PageKey, NewsItem>) {
+        val pageKey = params.key as PageKey.PageUrlKey
+
+        val result = fetchNewsItemsNextPage(pageKey.url)
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/rocket/content/news/data/dailyhunt/DailyHuntNewsRemoteDataSource.kt:42]: I'm so curious if we should run this code in a CoroutineScope? We can inject the CoroutineScope to this class. Then we don't have to rely on the background executor from the paging library and we can cancel the job by ourselves.

File: app/src/main/java/org/mozilla/rocket/content/news/ui/NewsFragment.kt [hunk 5]
```diff
@@ -77,40 +78,60 @@ class NewsFragment : Fragment() {
...
+                    is NewsUiModel -> item.componentId.hashCode().toLong()
+                    null -> super.getItemId(position)
+                    else -> error("unexpected type: ${item.javaClass}")
+                }
+            }
+        }.apply {
+            setHasStableIds(true)
+        }
 
-            news_list.monitorScrollImpression(telemetryViewModel)
```
Comment (benjamin-cheng) [inline app/src/main/java/org/mozilla/rocket/content/news/ui/NewsFragment.kt:104]: Please help to add the `monitorScrollImpression` back or we will lose the scroll impression telemetry.

File: app/src/main/java/org/mozilla/rocket/content/news/data/dailyhunt/DailyHuntNewsRemoteDataSource.kt (hunk 4 above)
Comment (proevan) [reply app/src/main/java/org/mozilla/rocket/content/news/data/dailyhunt/DailyHuntNewsRemoteDataSource.kt:42]: yeah, I created a ticket to mark this as a todo
https://github.com/mozilla-tw/FirefoxLite/issues/4928

Comment (proevan) [reply app/src/main/java/org/mozilla/rocket/content/news/data/dailyhunt/DailyHuntNewsRemoteDataSource.kt:42]: yeah, I created a ticket to mark this as a todo
#4928

File: app/src/main/java/org/mozilla/rocket/content/news/data/NewsDataSourceFactory.kt [hunk 6]
```diff
@@ -1,24 +1,35 @@
...
+    override fun create(): DataSource<PageKey, NewsItem> {
         val newsProvider = NewsProvider.getNewsProvider()
         return if (newsProvider?.isNewsPoint() == true) {
             val dailyHuntProvider = DailyHuntProvider.getProvider(appContext)
             if (dailyHuntProvider?.shouldEnable(appContext) == true) {
-                NewsRepository(DailyHuntNewsRemoteDataSource(appContext, dailyHuntProvider))
+                DailyHuntNewsRemoteDataSource(appContext, dailyHuntProvider, category, language)
             } else {
-                NewsRepository(NewsPointNewsRemoteDataSource(newsProvider))
+                NewsPointNewsRemoteDataSource(NewsProvider.getNewsProvider(), category, language)
```
[the remark this answers is not in the document]
Comment (proevan) [reply app/src/main/java/org/mozilla/rocket/content/news/data/NewsDataSourceFactory.kt:24]: fixed

File: app/src/main/java/org/mozilla/rocket/content/news/ui/NewsFragment.kt (hunk 5 above)
Comment (proevan) [reply app/src/main/java/org/mozilla/rocket/content/news/ui/NewsFragment.kt:104]: fixed, thanks!

**Pull request merged**